src/ccutil/errcode.h

```diff
@@ -37,18 +37,17 @@ enum TessErrorLogCode {
 #define MEMORY_ABORT 2
 #define FILE_ABORT 3
 
-#if !defined(__GNUC__) && !defined(__attribute__)
-# define __attribute__(attr) // compiler without support for __attribute__
-#endif
-
 class TESS_API ERRCODE { // error handler class
   const char *message;   // error message
 public:
+#if defined(__GNUC__) && defined(__attribute__)
```

Comment:
Is `__attribute__`defined anywhere? If not, then the new code simply disables all attributes.

Reply:
For gcc and regular clang? Are you sure?

`__ attribute__() `can be uses before the function declaration instead of in the end.

Reply:
Yes, but it is compiled conditionally, and the condition `&& defined(__attribute__)` is now always `false`.

Reply:
I thought `_attribute_` is auto defined by the compiler. Any way, the second condition can be dropped.

Reply:
>then the new code simply disables all attributes.

I don't think you are right.